utils/tfw_logger.cc

```diff
@@ -333,136 +357,269 @@ void stop_daemon()
-		 * to hold this lock, preventing subsequent executions of tempesta.sh.
-		 *
-		 * Close all descriptors before daemonizing.
-		 */
-		closefrom(3);
+std::shared_ptr<spdlog::logger> 
+setup_logger(const TfwLoggerConfig &config)
+{
+    try
+    {
+        // Create directory if it doesn't exist
+        fs::create_directories(fs::path(config.get_log_path()).parent_path());
+ 
+        auto logger = spdlog::basic_logger_mt("access_logger", config.get_log_path().string());
+        spdlog::set_default_logger(logger);
+        spdlog::set_level(config.get_debug() ? spdlog::level::debug : spdlog::level::info);
+        logger->flush_on(spdlog::level::info);
+        return logger;
+    }
+    catch (const spdlog::spdlog_ex &ex)
+    {
+        throw Except("Log initialization failed: {}", ex.what());
+    }
+}
 
-		logger = spdlog::basic_logger_mt("access_logger",
-						 vm["log"].as<std::string>());
-		spdlog::set_default_logger(logger);
-		spdlog::set_level(spdlog::level::info);
-		logger->flush_on(spdlog::level::info);
+size_t determine_cpu_count(const TfwLoggerConfig &config)
+{
+	size_t cpu_cnt;
+ 
+    if (config.get_cpu_count() > 0)
+    {
+        cpu_cnt = config.get_cpu_count();
+    }
+    else
+    {
+        cpu_cnt = std::thread::hardware_concurrency();
+        if (cpu_cnt == 0)
+        {
+            cpu_cnt = sysconf(_SC_NPROCESSORS_ONLN);
+            if (cpu_cnt < 0)
+                throw Except("Can't get CPU number");
+        }
+    }
+    spdlog::info("Using {} CPU(s)", cpu_cnt);
+    return cpu_cnt;
+}
 
-		spdlog::info("Starting daemon...");
+void 
+daemonize_if_needed(const TfwLoggerConfig &config)
+{
+	if (config.get_mode() == TfwLoggerConfig::Mode::DAEMON)
+	{
+		spdlog::info("Starting in daemon mode");
+
+		/*
+		* When the daemon forks, it inherits the file descriptor for
+		* /tmp/tempesta-lock-file, which was originally opened and locked by flock
+		* in the tempesta.sh script. After daemonizing, the daemon process continues
+		* to hold this lock, preventing subsequent executions of tempesta.sh.
+		*
+		* Close all descriptors before daemonizing.
+		*/
+		closefrom(3);
 
 		if (daemon(0, 0) < 0)
 			throw Except("Daemonization failed");
+ 
+		// Re-initialize logger after daemonizing
+		setup_logger(config);
+	}
+	else
+	{
+		spdlog::info("Starting in handle (foreground) mode");
+	}
+}
 
-		set_sig_handlers();
-
-		std::ofstream pid_file(pid_file_path);
-		if (!pid_file)
-			throw Except("Failed to open PID file");
-		pid_file << getpid();
-		pid_file.close();
-
-		while ((fd = open(dev_path, O_RDWR)) == -1) {
-			if (stop_flag.load(std::memory_order_acquire))
-				goto end;
-			if (errno != ENOENT)
-				throw Except("Can't open device");
-			std::this_thread::sleep_for(wait_for_dev);
+void 
+create_pid_file(const fs::path &pid_file_path)
+{
+	std::ofstream pid_file(pid_file_path);
+	if (!pid_file)
+		throw Except("Failed to open PID file");
+	pid_file << getpid();
+	pid_file.close();
+}
```

Comment:
what's the purpose of the pid file? Typically in Linux we also should lock it and set the right permissions. @ttaym you recently worked with a nice small library for PID files, could you please reference it  that we can borrow it's code here as well?